packages/preact/src/search/searchWithCachePrefill.ts

```diff
@@ -0,0 +1,99 @@
+import { search } from "@core/search"
+import { SearchOptions } from "@core/types"
+import { SearchQuery, SearchResult } from "@nosto/nosto-js/client"
+import { Config } from "@preact/config/config"
+import { deepMerge } from "@utils/deepMerge"
+
+import { applyQueryDefaults } from "./defaults"
+import { cacheSearchResult, loadCachedResultIfApplicable } from "./resultCaching"
+
+type CacheContextOptions = {
+  cacheResult?: SearchResult
+  prefillQuery?: SearchQuery
+}
+
+export async function searchWithCachePrefill(config: Config, searchQuery: SearchQuery, options: SearchOptions) {
+  const usePersistentCache = config.pageType !== "autocomplete" && config.persistentSearchCache
+
+  const cacheContextOptions = getCachePrefillOptions(usePersistentCache, searchQuery)
+
+  if (cacheContextOptions?.cacheResult) {
+    if (!cacheContextOptions?.prefillQuery) {
+      return cacheContextOptions.cacheResult
+    }
+    const backfillResponse = await performSearch(config, options, cacheContextOptions.prefillQuery)
+    const mergedResponse = {
+      ...cacheContextOptions.cacheResult,
+      products: {
+        ...cacheContextOptions.cacheResult.products,
+        hits: [...(cacheContextOptions.cacheResult.products?.hits || []), ...(backfillResponse.products?.hits || [])],
+        total: backfillResponse.products?.total
+      }
+    }
+    cacheSearchResult(usePersistentCache, searchQuery, mergedResponse)
+    return mergedResponse
+  }
+
+  const response = await performSearch(config, options, searchQuery)
+  cacheSearchResult(usePersistentCache, searchQuery, response)
+  return response
+}
+
+function getCachePrefillOptions(
+  usePersistentCache: boolean,
+  searchQuery: SearchQuery
+): CacheContextOptions | undefined {
+  const { from, size = 0 } = searchQuery.products || {}
+
+  const cacheData = loadCachedResultIfApplicable(usePersistentCache, searchQuery)
```

Comment:
does this logic require persistent caching?

Reply:
It's previous logic. I just retained it. It checks if the flag is used and returns null if it's not.